### PR TITLE
Fix TransformerEncoderLayer for bias=False

### PR DESCRIPTION
**Pull request by mikaylagawarecki**
Fixes https://github.com/pytorch/pytorch/issues/116385

Don't call `torch._transformer_encoder_layer_fwd` when `bias=False`

`bias=False` was not something that `torch._transformer_encoder_layer_fwd`  was meant to work with, it was my bad that this wasn't tested as I approved https://github.com/pytorch/pytorch/pull/101687.

`bias=False` was causing the `tensor_args` in [`TransformerEncoder`](https://github.com/pytorch/pytorch/blob/a17de2d6455e262f9b514584443ac60cf381bc85/torch/nn/modules/transformer.py#L364-L378)/[`TransformerEncoderLayer`](https://github.com/pytorch/pytorch/blob/a17de2d6455e262f9b514584443ac60cf381bc85/torch/nn/modules/transformer.py#L663-L677) to contain `None`s and error on checks for the fastpath like `t.requires_grad for t in tensor_args`.

Alternative fix would be to
1) Pass `torch.zeros_like({*}.weight)` to the kernel when `bias=False` and filter `tensor_args` as appropriate
2) Fix `torch._transformer_encoder_layer_fwd` to take `Optional<Tensor>` for biases and fix the kernels as appropriate

Let me know if these approaches are preferable

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #116760

### CODE REVIEW

File: torch/testing/_internal/common_modules.py [hunk 1]
```diff
@@ -2374,52 +2374,57 @@ def module_inputs_torch_nn_TransformerEncoderLayer(module_info, device, dtype, r
...
+                        4, 2, 8, dropout=0.0, batch_first=True, norm_first=norm_first, bias=bias
+                    ),
                     forward_input=FunctionInput(
                         make_input((2, 3, 4)),
                     ),
-                    reference_fn=fast_path_reference_fn,
-                    desc="fast_path_norm_first" if norm_first else "fast_path"
+                    # fastpath doesn't run when bias=False
+                    reference_fn=fast_path_reference_fn if bias else None,
+                    desc=f'fastpath_{bias}_norm_first_{norm_first}'
```
Comment (mikaylagawarecki) [inline torch/testing/_internal/common_modules.py:2427]: The existing fastpath test was not actually testing the fastpath against train mode since the fastpath is only run in `no_grad` mode and `TestModule.test_forward` is not run under `no_grad` context

I modified this such that the outputs in train mode are compared against a reference that runs the fastpath in eval/no_grad mode

File: torch/testing/_internal/common_modules.py [hunk 2]
```diff
@@ -2443,40 +2448,51 @@ def module_inputs_torch_nn_TransformerDecoderLayer(module_info, device, dtype, r
...
+                                                norm_first=norm_first, bias=bias),
                 forward_input=FunctionInput(
                     make_input((3, 4)), make_input((3, 4)), tgt_mask=tgt_mask, memory_mask=memory_mask,
                     tgt_key_padding_mask=tgt_key_padding_mask, memory_key_padding_mask=memory_key_padding_mask
                 ),
                 reference_fn=partial(no_batch_dim_reference_fn,
-                                     batch_first=True,
+                                     batch_first=batch_first,
                                      kwargs_to_batchify={'tgt_key_padding_mask': 0, 'memory_key_padding_mask': 0}),
-                desc='no_batch_dim_batch_first'
```
Comment (mikaylagawarecki) [inline torch/testing/_internal/common_modules.py:2466]: It feels like `batch_first` should not affect `no_batch_dim` so not entirely sure why we have these tests, but keeping them in case they are testing an edge case

Comment (jbschlosser) [reply torch/testing/_internal/common_modules.py:2466]: I do think it's a good black-box testing surface to cover. IIRC there was a problem at some point where some `batch_first` setting blew up on no-batch-dim inputs.

File: torch/testing/_internal/common_modules.py (hunk 1 above)
Comment (jbschlosser) [reply torch/testing/_internal/common_modules.py:2427]: > I modified this such that the outputs in train mode are compared against a reference that runs the fastpath in eval/no_grad mode

possibly dumb Q: do we need to compare train mode outputs vs. fastpath reference?

**Review by jbschlosser — APPROVED** [review]
Thanks! Fix looks solid, couple minor testing comments for `ModuleInfo` tests. As long as things pass, I'm good with it :)

File: torch/testing/_internal/common_modules.py (hunk 1 above)
Comment (mikaylagawarecki) [reply torch/testing/_internal/common_modules.py:2427]: from offline discussion: this swaps the reference to use fastpath in order to run it under no_grad mode since dropout=0 the outputs should match
